File: herald-admin-typescript/src/namespaces/members.ts
import type { HttpTransport } from "../transport.js";
import type { Member } from "../types.js";

export class MemberNamespace {
  constructor(private transport: HttpTransport) {}

  async add(streamId: string, userId: string, role?: string): Promise<Member> {
    return this.transport.request<Member>(
      "POST",
      `/streams/${encodeURIComponent(streamId)}/members`,
      { user_id: userId, role },
    );
  }

  async list(streamId: string): Promise<Member[]> {
    const resp = await this.transport.request<{ members: Member[] }>(
      "GET",
      `/streams/${encodeURIComponent(streamId)}/members`,
    );
    return resp.members;
  }

  async remove(streamId: string, userId: string): Promise<void> {
    return this.transport.request(
      "DELETE",
      `/streams/${encodeURIComponent(streamId)}/members/${encodeURIComponent(userId)}`,
    );
  }

  async update(streamId: string, userId: string, role: string): Promise<void> {
    return this.transport.request(
      "PATCH",
      `/streams/${encodeURIComponent(streamId)}/members/${encodeURIComponent(userId)}`,
      { role },
    );
  }
}

File: herald-admin-ruby/lib/herald_admin/members.rb
# frozen_string_literal: true

module HeraldAdmin
  class MemberNamespace
    def initialize(transport)
      @t = transport
    end

    def add(room_id, user_id, role: "member")
      data = @t.request("POST", "/rooms/#{esc(room_id)}/members", { user_id: user_id, role: role })
      Member.new(**data.transform_keys(&:to_sym))
    end

    def list(room_id)
      data = @t.request("GET", "/rooms/#{esc(room_id)}/members")
      data["members"].map { |m| Member.new(**m.transform_keys(&:to_sym)) }
    end

    def remove(room_id, user_id)
      @t.request("DELETE", "/rooms/#{esc(room_id)}/members/#{esc(user_id)}")
    end

    def update(room_id, user_id, role:)
      @t.request("PATCH", "/rooms/#{esc(room_id)}/members/#{esc(user_id)}", { role: role })
    end

    private

    def esc(s) = ERB::Util.url_encode(s)
  end
end
